flink-java/src/main/java/org/apache/flink/api/java/ScalaObjectChecker.java

```diff
@@ -0,0 +1,51 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.api.java;
+
+import org.apache.flink.annotation.Internal;
+import org.apache.flink.api.common.InvalidProgramException;
+
+/**
+ * Scala Object checker tries to verify if a class is implemented by
+ * Scala Object
+ */
+@Internal
+public class ScalaObjectChecker {
+	public static boolean isPotentialScalaObject(Object o) {
```

Comment:
The function you advised has a shortcoming, all functions in the call stack need to have this `(implicit ev: A <:< Singleton = null)` implicit parameter to make it work(see https://gist.github.com/Renkai/ddd59337d3c302eb3d061ff8d249413c),So that too many function signatures need to be changed, and the engineering complexity would be increased a lot. I'm afraid we have to keep the current version of ScalaObjectChecker until we found better solutions.